cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,567 @@
+    //     the ReplicaBasedBufferSpec will be used
+    //   - for GKE CCC or Karpenter NodePool the NodeClassBufferSpec will be used
+    //   - in case of no target object the user will be able to directly define the cluster capacity
+    //   - … - custom crd if the user runs their own buffer controller that can handle this target
+    TargetRef *TypedLocalObjectReference
+    Capacity  BufferCapacity
+
+    // Initially only active stand by meaning scaling up the cluster will be available.
+    // In the future the cloud providers will be able to offer their own buffer types.
+    Type string
+
+    // How the buffer should be maintained over time:
+    // - KeepEmpty - upscale the cluster to keep the empty buffer on top of the running workloads
+    // - KeepMinimum - upscale the cluster and let workloads fill in the buffer as needed
+    RecreationStrategy string
+}
+
+type BufferCapacity struct {
+    // One of the following must be set
+    // +optional
+    Replicas *ReplicaBufferCapacity
+    // +optional
+    NodeClass *NodeClassBufferCapacity
+    // +optional
+    Pods *PodBufferCapacity
+}
+
+type ReplicaBufferCapacity struct {
+    // One of the following must be set:
+    // +optional
+    Exactly *int
+    // +optional
+    Percent *PercentReplicaBufferCapacity
+}
+
+type PercentReplicaBufferCapacity struct {
+    Percent  int
+    MinCount *int
+    MaxCount *int
+}
+
+type NodeClassBufferCapacity struct {
+    // Upper limit of the total capacity that will be scaled for this buffer.
+    // The total capacity will be split into smaller chunks that each fits onto nodes.
+    // The size of a chunk will be either decided for the user (using some defaults and heuristics) or the
+    // user can specify per pod Requests to define the shape of the buffer space. In the latter case the
+    // CPU and Mem will be the maximum size of the buffer and if the Requests are too big to fit the total
+    // buffer size the admission of the buffer should be denied.
+    TotalCpu resource.Quantity
+    TotalMem resource.Quantity
+
+    // Optional if defining a single chunk size is needed. The controller will translate this into enough
+    // pods to create buffer not exceeding the total size above
+    PerChunkResources *ResourceLimits
+    // Optional if more detailed limitations than target object ref on its own is not enough.
+    NodeSelector map[string]string
+}
+
+// Subject to change depending on the sig-scheduling work to define workloads and reservations API
+type PodBufferCapacity struct {
+    ReplicaNum int
+    PodSpec    PodSpec
+}
+
+type BufferStatus struct {
+    PodBasedBufferSpec   PodBufferCapacity
+    Conditions           []BufferCondition
+    LastFullyProvisioned *metav1.Time
+}
+```
+
+### What is the responsibility split between the Buffers controller and autoscaler?
+
+The controller would own the logic of translating the buffer configuration into
```

Comment:
Could you rewrite this part more concretely, in terms of the new proposed objects? E.g. that autoscalers would only react to `BufferStatus.PodBasedBufferSpec` etc.

More generally, it's not clear from the proposal how the new objects are supposed to be interacted with by either a Node autoscaler or the Buffer controller. Maybe instead of having this section we should have separate sections explaining how both sides should use the API? Including which fields they consume and set, what kinds of objects the reference Buffer controller can translate, etc..